src/blockdata/script.rs

```diff
@@ -632,113 +632,79 @@ pub enum Instruction<'a> {
 
 /// Iterator over a script returning parsed opcodes
 pub struct Instructions<'a> {
-    data: &'a [u8],
+    data: ::core::slice::Iter<'a, u8>,
     enforce_minimal: bool,
 }
 
+impl<'a> Instructions<'a> {
+    fn next_push_data_len(&mut self, len: usize, max: usize) -> Option<Result<Instruction<'a>, Error>> {
+        let n = match read_uint_iter(&mut self.data, len) {
+            Ok(n) => n,
+            // We do exhaustive matching to not forget to handle new variants if we extend
+            // `UintError` type.
+            // Overflow actually means early end of script (script is definitely shorter
+            // than `usize::max_value()`)
+            Err(UintError::EarlyEndOfScript) | Err(UintError::NumericOverflow) => {
+                let data_len = self.data.len();
+                self.data.nth(data_len);  // Kill iterator so that it does not return an infinite stream of errors
+                return Some(Err(Error::EarlyEndOfScript));
+            },
+        };
+        if self.enforce_minimal && n < max {
+            let data_len = self.data.len();
+            self.data.nth(data_len);  // Kill iterator so that it does not return an infinite stream of errors
+            return Some(Err(Error::NonMinimalPush));
+        }
+        let ret = Some(Ok(Instruction::PushBytes(&self.data.as_slice()[..n])));
+        self.data.nth(n.max(1) - 1);
+        ret
+    }
+}
+
 impl<'a> Iterator for Instructions<'a> {
     type Item = Result<Instruction<'a>, Error>;
 
     fn next(&mut self) -> Option<Result<Instruction<'a>, Error>> {
```

Comment:
Maybe we can use this opportunity to rename the method into `try_next`?

Reply:
We can't rename `Iterator` methods.

Reply:
Sorry, my bad